packages/engine/lib/memory/src/shared_memory/segment.rs

```diff
@@ -181,10 +200,36 @@ impl Segment {
     /// reloading
     pub fn resize(&mut self, mut new_size: usize) -> Result<()> {
         new_size = Self::calculate_total_size(new_size, self.include_terminal_padding)?;
-        tracing::trace!("Trying to resize memory to: {}", new_size);
-        self.data.resize(new_size)?;
-        self.size = new_size;
-        Ok(())
+        tracing::trace!(
+            old_size = self.size,
+            new_size,
+            id = self.data.get_os_id(),
+            "Trying to resize memory"
+        );
+
+        if !cfg!(target_os = "macos") {
+            if self.data.resize(new_size).is_ok() {
+                self.size = new_size;
+                return Ok(());
+            } else {
+                tracing::warn!(
+                    old_size = self.size,
+                    new_size,
+                    id = self.data.get_os_id(),
+                    "Could not resize memory, using fallback"
+                );
+            }
+        }
+
+        let memory_id = MemoryId::from_os_id(self.data.get_os_id())?;
+        let new_data = ShmemConf::new(true)
+            .os_id(MemoryId::new(memory_id.base_id()).to_string())
+            .size(new_size)
+            .create()?;
+        unsafe { std::ptr::copy_nonoverlapping(self.data.as_ptr(), new_data.as_ptr(), self.size) };
```

Comment:
Can you please give an in-depth reasoning of why this works. I think there's some complex reasoning we're going to have to think about here with batch IDs and how the language runners load the batches, if they cache them, etc.

Having a new memory ID seems risky to me for the reasons stated above, but perhaps it's okay because of some other implications of this only happening on a resize, etc. 

Can you also give me instructions for this that are guaranteed to cause a resize so I can test this on macOS for you?

Reply:
Yes. The memory id is an implementation detail inside of `Segment` and it's not possible to read from memory directly without going through `Segment`'s implementation (This was the main change in #381). It's also not possible to store a `Segment` in an engine component and resize it at any other place because `resize` requires a mutable `Segment`. This also implies that the runners cannot unmap memory, as they only have access to the segment, which only gives access to a slice of memory. After creating a `Segment` a memory id is completely opaque. 
In fact we never used the implementation of `Segment`, so I added 669b9ca4d156e92b283cc42dbd0048950f5b5449, to ensure this at compile time as well.